### PR TITLE
Introduce Dogecoin block subsidy calculations

### PR DESCRIPTION
**Pull request by rnicoll**

### CODE REVIEW

File: src/dogecoin.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        // This extracts 7*4 bits out of the middle of the previous block hash to use as a seed.
+        // It's longer than the 1.14 code, but avoids the locale-sensitive strtol() function.
+        const int bitsPerNibble = 4;
+        const int totalNibbles = 256 / bitsPerNibble;
+        const int nibblesToClear = 7;
+        const int nibblesToKeep = 7;
+        arith_uint256 upperBits = UintToArith256(prevHash) << (nibblesToClear * bitsPerNibble);
+        arith_uint256 cleanedBits = upperBits >> ((totalNibbles - nibblesToKeep) * bitsPerNibble);
+        const std::string cseed_str = prevHash.ToString().substr(7, 7);
```
Comment (rnicoll) [inline src/dogecoin.cpp:34]:
```suggestion
```

File: src/miner.cpp [hunk 2]
```diff
@@ -160,7 +161,8 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& sc
     coinbaseTx.vin[0].prevout.SetNull();
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = scriptPubKeyIn;
-    coinbaseTx.vout[0].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus());
+    coinbaseTx.vout[0].nValue = nFees + GetDogecoinBlockSubsidy(nHeight, chainparams.GetConsensus(), pindexPrev->GetBlockHash(
+));
```
Comment (patricklodder) [inline src/miner.cpp:165]: newline why?

File: src/test/dogecoin_tests.cpp [hunk 3]
```diff
@@ -12,6 +12,108 @@
 
 BOOST_FIXTURE_TEST_SUITE(dogecoin_tests, TestingSetup)
 
+/**
+ * the maximum block reward at a given height for a block without fees
+ */
+uint64_t expectedMaxSubsidy(int height) {
```
Comment (patricklodder) [inline src/test/dogecoin_tests.cpp:18]: where is this (and `expectedMinSubsidy`) used?

File: src/wallet/test/wallet_tests.cpp [hunk 4]
```diff
@@ -339,7 +340,8 @@ BOOST_FIXTURE_TEST_CASE(coin_mark_dirty_immature_credit, TestChain240Setup)
     // credit amount is calculated.
     wtx.MarkDirty();
     BOOST_CHECK(spk_man->AddKeyPubKey(coinbaseKey, coinbaseKey.GetPubKey()));
-    BOOST_CHECK_EQUAL(wtx.GetImmatureCredit(), 25*COIN);
+    // Dogecoin: Scaled up by 1,000
```
Comment (patricklodder) [inline src/wallet/test/wallet_tests.cpp:343]: scaled up by 10,000 in the code?

File: src/wallet/test/wallet_tests.cpp [hunk 5]
```diff
@@ -147,7 +147,8 @@ BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain240Setup)
         BOOST_CHECK_EQUAL(result.last_failed_block, oldTip->GetBlockHash());
         BOOST_CHECK_EQUAL(result.last_scanned_block, newTip->GetBlockHash());
         BOOST_CHECK_EQUAL(*result.last_scanned_height, newTip->nHeight);
-        BOOST_CHECK_EQUAL(wallet.GetBalance().m_mine_immature, 25 * COIN);
+	// Dogecoin: Scaled up by 1,000
```
Comment (patricklodder) [inline src/wallet/test/wallet_tests.cpp:150]: scaled up by 10,000 in the code?

**Review by patricklodder — COMMENTED** [review]
I have only nits as there is no point in comparing too deeply without scrypt.

File: src/test/dogecoin_tests.cpp (hunk 3 above)
Comment (rnicoll) [reply src/test/dogecoin_tests.cpp:18]: They're not, it turns out, it's a mistake copied from 1.14. Removing.

**Review by patricklodder — APPROVED** [review]
utACK. will start testing in-depth after merging scrypt.

**Pull request merged**